src/11_render_passes.rs

```diff
@@ -0,0 +1,486 @@
+extern crate env_logger;
+#[cfg(feature = "dx12")]
+extern crate gfx_backend_dx12 as back;
+#[cfg(feature = "metal")]
+extern crate gfx_backend_metal as back;
+#[cfg(feature = "vulkan")]
+extern crate gfx_backend_vulkan as back;
+extern crate gfx_hal as hal;
+extern crate winit;
+#[macro_use]
+extern crate lazy_static;
+extern crate glsl_to_spirv;
+
+use hal::{Capability, Device, Instance, PhysicalDevice, QueueFamily, Surface, SwapchainConfig};
+use std::{fs, io::Read};
+
+static WINDOW_NAME: &str = "11_render_passes";
+
+fn main() {
+    log_output_setup();
+    let (window, events_loop) = init_window();
+    // we must return surface, since it must be destroyed before swapchain
+    let (
+        _instance,
+        device,
+        swapchain,
+        _surface,
+        frame_images,
+        render_pass,
+        descriptor_set_layouts,
+        pipeline_layout,
+    ) = init_hal(&window);
+    main_loop(events_loop);
+    // must destroy the swapchain
+    clean_up(
+        device,
+        frame_images,
+        swapchain,
+        render_pass,
+        descriptor_set_layouts,
+        pipeline_layout,
+    );
+}
+
+fn create_render_pass(
+    device: &<back::Backend as hal::Backend>::Device,
+    format: Option<hal::format::Format>,
+) -> <back::Backend as hal::Backend>::RenderPass {
+    let samples: u8 = 1;
+
+    let ops = hal::pass::AttachmentOps {
+        load: hal::pass::AttachmentLoadOp::Clear,
+        store: hal::pass::AttachmentStoreOp::Store,
+    };
+
+    let stencil_ops = hal::pass::AttachmentOps::DONT_CARE;
+
+    // TODO: give an explanation for why a half-open range works here...
+    // shouldn't we just have a 2-ple? (i::Layout::Undefined, i::Layout::Present)
```

Comment:
It's just to make it explicit that it will transition from `initial_layout` to `final_layout`, so `initial_layout..final_layout`. A tuple could have been used instead, but the range might be slightly clearer